apps/rabbitmq_prelaunch/src/rabbit_boot_state_systemd.erl

```diff
@@ -0,0 +1,160 @@
+%%%-------------------------------------------------------------------
+%% The contents of this file are subject to the Mozilla Public License
+%% Version 1.1 (the "License"); you may not use this file except in
+%% compliance with the License. You may obtain a copy of the License
+%% at https://www.mozilla.org/MPL/
+%%
+%% Software distributed under the License is distributed on an "AS IS"
+%% basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
+%% the License for the specific language governing rights and
+%% limitations under the License.
+%%
+%% The Original Code is RabbitMQ.
+%%
+%% The Initial Developer of the Original Code is GoPivotal, Inc.
+%% Copyright (c) 2007-2020 Pivotal Software, Inc.  All rights reserved.
+%%
+
+-module(rabbit_boot_state_systemd).
+
+-behavior(gen_server).
+
+-export([start_link/0]).
+
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2,
+         code_change/3]).
+
+-record(state, {sd_notify_module,
+                socket}).
+
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+init([]) ->
+    case code:load_file(sd_notify) of
+        {module, sd_notify} ->
+            {ok, #state{sd_notify_module = sd_notify}};
+        {error, _} ->
+            case rabbit_prelaunch:get_context() of
+                #{systemd_notify_socket := Socket} when Socket =/= undefined ->
+                    {ok, #state{socket = Socket}};
+                _ ->
+                    ignore
+            end
+    end.
+
+handle_call(_Request, _From, State) ->
+    {noreply, State}.
+
+handle_cast({notify_boot_state, ready}, #{sd_notify_module := SDNotify} = State) ->
+    rabbit_log_prelaunch:debug("notifying systemd of readiness via native module"),
+    sd_notify_legacy(SDNotify),
+    {noreply, State};
+handle_cast({notify_boot_state, ready}, #{socket := Socket} = State) ->
+    rabbit_log_prelaunch:debug("notifying systemd of readiness via socat"),
+    sd_notify_socat(Socket),
+    {noreply, State};
+handle_cast({notify_boot_state, _BootState}, State) ->
+    {noreply, State}.
+
+handle_info(Msg, State) ->
+    io:format(standard_error, "Unexpected message: ~p~n",[Msg]),
+    {noreply, State}.
+
+terminate(normal, _State) ->
+    io:format(standard_error, "~p terminating.~n", [?MODULE]),
```

Comment:
This one is not needed, except if you want to log more details about the termination. Erlang can take care of logging this kind of event if we enable it.

Reply:
I didn't realize that. Removed.